tests/test_config.py

```diff
@@ -183,6 +184,123 @@ def test_get(self):
                 self.assertIsInstance(server_config.auth, tuple)
 
 
+class ReprTestCase(TestCase):
+    """Test method ``nailgun.config.BaseServerConfig.__repr__``."""
+
+    def test_bsc_v1(self):
+        """Test :class:`nailgun.config.BaseServerConfig`.
+
+        Assert that ``__repr__`` works correctly when ``url`` is specified.
+
+        """
+        target = "nailgun.config.BaseServerConfig(url='bogus')"
+        self.assertEqual(target, repr(BaseServerConfig('bogus')))
+        import nailgun  # noqa pylint:disable=unused-variable
+        # pylint:disable=eval-used
+        self.assertEqual(target, repr(eval(repr(BaseServerConfig('bogus')))))
```

Comment:
> This is asserting that you can get the representation and just paste it in a python shell for example?

Yes. That is exactly correct.

> Also I'm interested more on the reason to do this kind of assertion :smile:

The job of both `__repr__` and `__str__` is to provide a string representation of an object. However, they provide different kinds of string representations. From the docs:

> In short, the goal of `__repr__` is to be unambiguous and `__str__` is to be readable.

`__repr__` should be *so* unambiguous that you can copy-and-paste its output in to a shell and get the same object back. From the docs for [`__repr__`](https://docs.python.org/3/reference/datamodel.html#object.__repr__):

> If at all possible, this should look like a valid Python expression that could be used to recreate an object with the same value (given an appropriate environment).

Here's another example of `__repr__` and `__str__`:

```python
>>> import datetime
>>> today = datetime.datetime.now()
>>> str(today)
'2012-03-14 09:21:58.130922'
>>> repr(today)
'datetime.datetime(2012, 3, 14, 9, 21, 58, 130922)'
```

As you can see, `__repr__` provides an "official" representation of an object that can be used to recreate a new object with the same value.

I've decided to implement `__repr__` instead of `_-str__` because the former is usable in a wider variety of situations and because it will be more useful for debugging. Check this out:

one important distinction between `__str__` and `__repr__` is what happens when a container is stringified. Check this out:

```python
>>> class Foo(object):
...   def __str__(self):
...     return 'Yellow Submarine'
... 
>>> str(Foo())
'Yellow Submarine'
>>> repr(Foo())
'<__main__.Foo object at 0x7f9ba3f86438>'
>>> str([Foo(), Foo()])
'[<__main__.Foo object at 0x7f9ba3f86438>, <__main__.Foo object at 0x7f9ba3f864a8>]'
```

What happens above is that Python calls `__repr__` on all of the elements of a container even when the container itself defines `__str__`.